clickhouse/base/socket.h

```diff
@@ -57,13 +59,31 @@ class windowsErrorCategory : public std::error_category {
 
 #endif
 
-class Socket {
+
+class SocketBase {
+public:
+    virtual ~SocketBase();
+
+    virtual std::unique_ptr<InputStream> makeInputStream() const = 0;
+    virtual std::unique_ptr<OutputStream> makeOutputStream() const = 0;
+};
+
+
+class SocketFactory {
```

Comment:
To simplify issue with reconnecting and timeouts (and to provide a framework for upcoming multiple-hosts #139 ). Please consider updating the interface:
```
class SocketFactory {
        public:
    virtual ~SocketFactory();

    // TODO: move connection-related options to ConnectionOptions structure.

    virtual std::unique_ptr<SocketBase> Connect(const NetworkAddress &, const ClientOptions &) = 0;
    // reconnects to a given address, maybe with a timeout or with any special procedures.
    virtual std::unique_ptr<SocketBase> ReConnect(const NetworkAddress &, const ClientOptions &) = 0;
};
```

Reply:
There's a problem with `NetworkAddress` - its constructor uses blocking `getaddrinfo`; one way of soling this is to make some base class for `NetworkAddress`, but i feel like this would complicate the interface too much (and we already have some Java-style abstract factories here)

Introducing `ConnectionOptions` seems like a good enhancement for me, but i believe it has to be done in #139, or at least on the receiving end of upcoming merge conflict. Since `SocketFactory` is a new feature, there shouldn't a concern with backwards compatibility